File: backend/app/models/reference_build.py
import uuid
from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.base import Base


class ReferenceBuild(Base):
    __tablename__ = "reference_builds"

    id          = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    build_key   = Column(String(64), nullable=False, unique=True, index=True)
    label       = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    total_approx = Column(Integer, nullable=False)
    is_active   = Column(Boolean, nullable=False, server_default="true")
    created_at  = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at  = Column(DateTime(timezone=True), nullable=False,
                         server_default=func.now(), onupdate=func.now())

    parts = relationship(
        "ReferenceBuildPart", back_populates="build",
        cascade="all, delete-orphan", order_by="ReferenceBuildPart.sort_order",
    )
class ReferenceBuildPart(Base):
    __tablename__ = "reference_build_parts"

    id           = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    build_id     = Column(UUID(as_uuid=True), ForeignKey("reference_builds.id", ondelete="CASCADE"),
                          nullable=False, index=True)
    part_id      = Column(UUID(as_uuid=True), ForeignKey("pc_parts.id", ondelete="RESTRICT"),
                          nullable=False, index=True)
    component    = Column(String(50), nullable=False)
    sort_order   = Column(Integer, nullable=False, default=0)
    approx_price = Column(Integer, nullable=False)

    build = relationship("ReferenceBuild", back_populates="parts")
    part  = relationship("PCPart")

File: admin/lib/admin/schema/reference_build.ex
defmodule Admin.Schema.ReferenceBuild do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, Ecto.UUID, autogenerate: false}

  schema "reference_builds" do
    field :build_key, :string
    field :label, :string
    field :description, :string
    field :total_approx, :integer
    field :is_active, :boolean, default: true

    timestamps(type: :utc_datetime, inserted_at: :created_at, updated_at: :updated_at)
  end

  def changeset(build, attrs, _metadata \\ %{}) do
    build
    |> cast(attrs, [:build_key, :label, :description, :total_approx, :is_active])
    |> validate_required([:build_key, :label, :description])
    |> unique_constraint(:build_key)
  end
end
